objects/src/assets/mod.rs

```diff
@@ -0,0 +1,420 @@
+
+    fn try_from(value: Word) -> Result<Self, Self::Error> {
+        let first_bit = value[0].as_int() >> 63;
+        match first_bit {
+            0 => NonFungibleAsset::try_from(value).map(Asset::from),
+            1 => FungibleAsset::try_from(value).map(Asset::from),
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl TryFrom<[u8; 32]> for Asset {
+    type Error = AssetError;
+
+    fn try_from(value: [u8; 32]) -> Result<Self, Self::Error> {
+        let first_bit = value[0] >> 7;
+        match first_bit {
+            0 => NonFungibleAsset::try_from(value).map(Asset::from),
+            1 => FungibleAsset::try_from(value).map(Asset::from),
+            _ => unreachable!(),
+        }
+    }
+}
+
+// FUNGIBLE ASSET
+// ================================================================================================
+/// A fungible asset.
+///
+/// A fungible asset consists of a faucet ID of the faucet which issued the asset as well as the
+/// asset amount. Asset amount is guaranteed to be 2^63 - 1 or smaller.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct FungibleAsset {
+    faucet_id: AccountId,
+    amount: u64,
+}
+
+impl FungibleAsset {
+    // CONSTANTS
+    // --------------------------------------------------------------------------------------------
+    /// Specifies a maximum amount value for fungible assets which can be at most a 63-bit value.
+    pub const MAX_AMOUNT: u64 = (1u64 << 63) - 1;
+
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    /// Returns a fungible asset instantiated with the provided faucet ID and amount.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - The provided faucet ID is not for a fungible asset faucet.
+    /// - The provide amount is greater than 2^63 - 1.
+    pub fn new(faucet_id: AccountId, amount: u64) -> Result<Self, AssetError> {
+        if !faucet_id.is_fungible_faucet() {
+            return Err(AssetError::not_a_fungible_faucet_id(faucet_id));
+        }
+
+        if amount > Self::MAX_AMOUNT {
+            return Err(AssetError::amount_too_big(amount));
+        }
+
+        Ok(Self { faucet_id, amount })
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Return ID of the faucet which issued this asset.
+    pub fn faucet_id(&self) -> AccountId {
+        self.faucet_id
+    }
+
+    /// Returns the amount of this asset.
+    pub fn amount(&self) -> u64 {
+        self.amount
+    }
+
+    // OPERATIONS
+    // --------------------------------------------------------------------------------------------
+
+    /// Adds two fungible assets together and returns the result.
+    ///
+    /// # Errors
+    /// Returns an error if the assets were not issued by the same faucet.
+    #[allow(clippy::should_implement_trait)]
+    pub fn add(self, other: Self) -> Result<Self, AssetError> {
```

Comment:
Should we consider using the stdlib add? I understand this function returns a result instead of the added value, but we could do something like this

```rust
impl core::ops::Add<&FungibleAsset> for FungibleAsset {
    type Output = Result<FungibleAsset, AssetError>;

    fn add(self, rhs: &FungibleAsset) -> Self::Output {
        if self.faucet_id != other.faucet_id {
            return Err(AssetError::some_divergent_asset_err());
        }
        self.amount.checked_add(rhs.amount).ok_or_else(|| AssetError::some_overflow_report(self, other))
    }
}
```

However, since this is actually a checked_add, it might make more sense to rename the method. Usually `add` is unchecked (overflow, etc)

Reply:
I thought about it too - but I don't know if we need this especially since we don't override any other operators and overriding just addition will look a bit weird, in my opinion.